added the description of the reading order of the config.h files. (#8545)
qmk/qmk_firmware: docs/hardware_keyboard_guidelines.md

All projects need to have a `config.h` file that sets things like the matrix size, product name, USB VID/PID, description and other settings. In general, use this file to set essential information and defaults for your keyboard that will always work.

The `config.h` files can also be placed in sub-folders, and the order in which they are read is as follows

* keuyboards/top_folder/config.h
  * keuyboards/top_folder/sub_1/config.h
    * keuyboards/top_folder/sub_1/sub_2/config.h
      * keuyboards/top_folder/sub_1/sub_2/sub_3/config.h
        * keuyboards/top_folder/sub_1/sub_2/sub_3/sub_4/config.h
          * users/an_user_folder/config.h
          * keuyboards/top_folder/keymaps/an_keymap/config.h
        * keuyboards/top_folder/sub_1/sub_2/sub_3/sub_4/post_config.h
      * keuyboards/top_folder/sub_1/sub_2/sub_3/post_config.h
    * keuyboards/top_folder/sub_1/sub_2/post_config.h
  * keuyboards/top_folder/sub_1/post_config.h

> * keuyboards/top_folder/post_config.h

Reviewer: ```suggestion
* keyboards/top_folder/config.h
  * keyboards/top_folder/sub_1/config.h
    * keyboards/top_folder/sub_1/sub_2/config.h
      * keyboards/top_folder/sub_1/sub_2/sub_3/config.h
        * keyboards/top_folder/sub_1/sub_2/sub_3/sub_4/config.h
          * users/a_user_folder/config.h
          * keyboards/top_folder/keymaps/a_keymap/config.h
        * keyboards/top_folder/sub_1/sub_2/sub_3/sub_4/post_config.h
      * keyboards/top_folder/sub_1/sub_2/sub_3/post_config.h
    * keyboards/top_folder/sub_1/sub_2/post_config.h
  * keyboards/top_folder/sub_1/post_config.h
* keyboards/top_folder/post_config.h
```

All projects need to have a `config.h` file that sets things like the matrix size, product name, USB VID/PID, description and other settings. In general, use this file to set essential information and defaults for your keyboard that will always work.

> The `config.h` files can also be placed in sub-folders, and the order in which they are read is as follows

Reviewer 2: ```suggestion
The `config.h` files can also be placed in sub-folders, and the order in which they are read is as follows:
```

* `keyboards/top_folder/config.h`
  * `keyboards/top_folder/sub_1/config.h`
    * `keyboards/top_folder/sub_1/sub_2/config.h`
      * `keyboards/top_folder/sub_1/sub_2/sub_3/config.h`
        * `keyboards/top_folder/sub_1/sub_2/sub_3/sub_4/config.h`
          * `users/a_user_folder/config.h`
          * `keyboards/top_folder/keymaps/a_keymap/config.h`
        * `keyboards/top_folder/sub_1/sub_2/sub_3/sub_4/post_config.h`
      * `keyboards/top_folder/sub_1/sub_2/sub_3/post_config.h`
    * `keyboards/top_folder/sub_1/sub_2/post_config.h`
  * `keyboards/top_folder/sub_1/post_config.h`
* `keyboards/top_folder/post_config.h`

The `post_config.h` file is for additional post-processing depending on what is specified in the `config.h` file.

> For example, if you define the `IOS_DEVICE_ENABLE` macro in the `keymaps/a_keymap/config.h` file as follows, you can configure the detailed settings accordingly in the `post_config.h` file.

Reviewer 2: ```suggestion
The `post_config.h` file can be used for additional post-processing, depending on what is specified in the `config.h` file. For example, if you define the `IOS_DEVICE_ENABLE` macro in your keymap-level `config.h` file as follows, you can configure more detailed settings accordingly in the `post_config.h` file:
```

* `keyboards/top_folder/keymaps/a_keymap/config.h`
  ```c
  #define IOS_DEVICE_ENABLE
  ```
* `keyboards/top_folder/post_config.h`
  ```c
  #ifndef IOS_DEVICE_ENABLE
    // USB_MAX_POWER_CONSUMPTION value for this keyboard
    #define USB_MAX_POWER_CONSUMPTION 400
  #else
    // fix iPhone and iPad power adapter issue
    // iOS device need lessthan 100
    #define USB_MAX_POWER_CONSUMPTION 100
  #endif
  
  #ifdef RGBLIGHT_ENABLE
    #ifndef IOS_DEVICE_ENABLE
      #define RGBLIGHT_LIMIT_VAL 200
      #define RGBLIGHT_VAL_STEP 17
    #else
      #define RGBLIGHT_LIMIT_VAL 35
      #define RGBLIGHT_VAL_STEP 4
    #endif
    #define RGBLIGHT_HUE_STEP 10

>     #define RGBLIGHT_SAT_STEP 17

Reviewer 2: This stuff may not work if the keyboard-level config.h already defines these.

Author: The definitions have been revised and a note added.

      #define RGBLIGHT_LIMIT_VAL 35
      #define RGBLIGHT_VAL_STEP 4
    #endif
    #ifndef RGBLIGHT_HUE_STEP
      #define RGBLIGHT_HUE_STEP 10
    #endif
    #ifndef RGBLIGHT_SAT_STEP
      #define RGBLIGHT_SAT_STEP 17
    #endif
  #endif
  ```

> NOTE: If you define `post_config.h` in the above example, you should not define USB_MAX_POWER_CONSUMPTION, RGBLIGHT_LIMIT_VAL, RGBLIGHT_VAL_STEP, RGBLIGHT_HUE_STEP and RGBLIGHT_SAT_STEP in keyboard-level `config.h`. And, the user-level `config.h` of this keyboard should not define USB_MAX_POWER_CONSUMPTION, RGBLIGHT_LIMIT_VAL, and RGBLIGHT_VAL_STEP.

Reviewer 2: ```suggestion
?> If you define options using `post_config.h` as in the above example, you should not define the same options in the keyboard- or user-level `config.h`.
```

Author: Thanks!
